In [1]:
import pandas as pd
import numpy as np

df_products = pd.read_csv("customer_products.csv")

print("Dataset shape:", df_products.shape)
display(df_products.head())

Dataset shape: (29965, 7)


,Customer_ID,Savings_Account,Credit_Card,Debit_Card,Insurance,Home_Loan,Personal_Loan
0,1,1,0,1,0,0,1
1,2,0,0,0,0,0,0
2,3,0,1,1,0,0,0
3,4,1,0,1,1,0,0
4,5,1,0,1,0,0,0


In [2]:
import pandas as pd
import numpy as np

df_products = pd.read_csv("customer_products.csv")

print("Dataset shape:", df_products.shape)
display(df_products.head())

Dataset shape: (29965, 7)


,Customer_ID,Savings_Account,Credit_Card,Debit_Card,Insurance,Home_Loan,Personal_Loan
0,1,1,0,1,0,0,1
1,2,0,0,0,0,0,0
2,3,0,1,1,0,0,0
3,4,1,0,1,1,0,0
4,5,1,0,1,0,0,0


In [3]:
product_columns = [
    "Savings_Account",
    "Credit_Card",
    "Debit_Card",
    "Insurance",
    "Home_Loan",
    "Personal_Loan"
]

transactions_df = df_products[product_columns].astype(bool)

transactions_df = transactions_df[
    transactions_df.sum(axis=1) > 0
].copy()

print("Valid customer transactions:", len(transactions_df))

Valid customer transactions: 29290


from mlxtend.frequent_patterns import apriori

In [5]:
!pip install mlxtend

   ---------------------------------------- 0.0/1.4 MB ? eta -:--:--
   ----------------------- ---------------- 0.8/1.4 MB 16.9 MB/s eta 0:00:01
   ---------------------------------------- 1.4/1.4 MB 4.7 MB/s  0:00:00


In [6]:
from mlxtend.frequent_patterns import apriori

In [7]:
frequent_itemsets = apriori(
    transactions_df,
    min_support=0.10,
    use_colnames=True
)

In [8]:
print("Number of frequent itemsets:", len(frequent_itemsets))

display(frequent_itemsets)

Number of frequent itemsets: 19


,support,itemsets
0,0.719017,frozenset({Savings_Account})
1,0.358484,frozenset({Credit_Card})
2,0.764493,frozenset({Debit_Card})
3,0.306589,frozenset({Insurance})
4,0.155411,frozenset({Home_Loan})
5,0.251895,frozenset({Personal_Loan})
6,0.253602,"frozenset({Credit_Card, Savings_Account})"
7,0.538409,"frozenset({Debit_Card, Savings_Account})"
8,0.215193,"frozenset({Insurance, Savings_Account})"
9,0.108365,"frozenset({Home_Loan, Savings_Account})"


In [9]:
frequent_itemsets = frequent_itemsets.sort_values(
    by="support",
    ascending=False
).reset_index(drop=True)

display(frequent_itemsets)

,support,itemsets
0,0.764493,frozenset({Debit_Card})
1,0.719017,frozenset({Savings_Account})
2,0.538409,"frozenset({Debit_Card, Savings_Account})"
3,0.358484,frozenset({Credit_Card})
4,0.306589,frozenset({Insurance})
5,0.267327,"frozenset({Credit_Card, Debit_Card})"
6,0.253602,"frozenset({Credit_Card, Savings_Account})"
7,0.251895,frozenset({Personal_Loan})
8,0.229396,"frozenset({Insurance, Debit_Card})"
9,0.215193,"frozenset({Insurance, Savings_Account})"


In [10]:
frequent_itemsets["itemsets"] = frequent_itemsets["itemsets"].apply(
    lambda x: ", ".join(sorted(x))
)

display(frequent_itemsets)

,support,itemsets
0,0.764493,Debit_Card
1,0.719017,Savings_Account
2,0.538409,"Debit_Card, Savings_Account"
3,0.358484,Credit_Card
4,0.306589,Insurance
5,0.267327,"Credit_Card, Debit_Card"
6,0.253602,"Credit_Card, Savings_Account"
7,0.251895,Personal_Loan
8,0.229396,"Debit_Card, Insurance"
9,0.215193,"Insurance, Savings_Account"


In [11]:
min_support=0.10
print("Number of frequent itemsets:", len(frequent_itemsets))
display(frequent_itemsets)

Number of frequent itemsets: 19


,support,itemsets
0,0.764493,Debit_Card
1,0.719017,Savings_Account
2,0.538409,"Debit_Card, Savings_Account"
3,0.358484,Credit_Card
4,0.306589,Insurance
5,0.267327,"Credit_Card, Debit_Card"
6,0.253602,"Credit_Card, Savings_Account"
7,0.251895,Personal_Loan
8,0.229396,"Debit_Card, Insurance"
9,0.215193,"Insurance, Savings_Account"


In [13]:
from mlxtend.frequent_patterns import apriori

# Recreate frequent itemsets in the ORIGINAL format
frequent_itemsets = apriori(
    transactions_df,
    min_support=0.10,
    use_colnames=True
)

print("Number of frequent itemsets:", len(frequent_itemsets))

display(frequent_itemsets)

Number of frequent itemsets: 19


,support,itemsets
0,0.719017,frozenset({Savings_Account})
1,0.358484,frozenset({Credit_Card})
2,0.764493,frozenset({Debit_Card})
3,0.306589,frozenset({Insurance})
4,0.155411,frozenset({Home_Loan})
5,0.251895,frozenset({Personal_Loan})
6,0.253602,"frozenset({Credit_Card, Savings_Account})"
7,0.538409,"frozenset({Debit_Card, Savings_Account})"
8,0.215193,"frozenset({Insurance, Savings_Account})"
9,0.108365,"frozenset({Home_Loan, Savings_Account})"


In [14]:
from mlxtend.frequent_patterns import association_rules

rules = association_rules(
    frequent_itemsets,
    metric="confidence",
    min_threshold=0.50
)

print("Number of association rules:", len(rules))

Number of association rules: 19


In [15]:
rules_display = rules[
    [
        "antecedents",
        "consequents",
        "support",
        "confidence",
        "lift"
    ]
].copy()

rules_display["antecedents"] = rules_display["antecedents"].apply(
    lambda x: ", ".join(sorted(x))
)

rules_display["consequents"] = rules_display["consequents"].apply(
    lambda x: ", ".join(sorted(x))
)

rules_display = rules_display.sort_values(
    by="lift",
    ascending=False
).reset_index(drop=True)

display(rules_display)

,antecedents,consequents,support,confidence,lift
0,"Credit_Card, Debit_Card",Savings_Account,0.189245,0.707918,0.984564
1,Credit_Card,Savings_Account,0.253602,0.707429,0.983883
2,Personal_Loan,"Debit_Card, Savings_Account",0.133424,0.529683,0.983793
3,"Personal_Loan, Savings_Account",Debit_Card,0.133424,0.751683,0.983244
4,"Debit_Card, Personal_Loan",Savings_Account,0.133424,0.706946,0.983213
5,"Insurance, Savings_Account",Debit_Card,0.161591,0.750912,0.982236
6,Credit_Card,"Debit_Card, Savings_Account",0.189245,0.527905,0.980490
7,Personal_Loan,Debit_Card,0.188733,0.749255,0.980067
8,Personal_Loan,Savings_Account,0.177501,0.704663,0.980036
9,"Debit_Card, Insurance",Savings_Account,0.161591,0.704420,0.979699


In [16]:
import numpy as np
import pandas as pd

np.random.seed(42)

n = len(df_products)

df_products_v2 = pd.DataFrame()

df_products_v2["Customer_ID"] = range(1, n + 1)

In [17]:
df_products_v2["Savings_Account"] = np.random.binomial(
    1, 0.70, n
)

In [18]:
debit_prob = np.where(
    df_products_v2["Savings_Account"] == 1,
    0.90,
    0.35
)

df_products_v2["Debit_Card"] = np.random.binomial(
    1, debit_prob
)

In [19]:
credit_prob = np.where(
    (df_products_v2["Savings_Account"] == 1) &
    (df_products_v2["Debit_Card"] == 1),
    0.60,
    0.15
)

df_products_v2["Credit_Card"] = np.random.binomial(
    1, credit_prob
)

In [20]:
insurance_prob = np.where(
    (df_products_v2["Home_Loan"] if "Home_Loan" in df_products_v2 else 0) == 1,
    0.75,
    0.25
)

In [21]:
df_products_v2["Home_Loan"] = np.random.binomial(
    1, 0.15, n
)
insurance_prob = np.where(
    df_products_v2["Home_Loan"] == 1,
    0.75,
    0.20
)

df_products_v2["Insurance"] = np.random.binomial(
    1, insurance_prob
)

In [22]:
df_products_v2["Personal_Loan"] = np.random.binomial(
    1, 0.25, n
)
insurance_prob = np.where(
    (df_products_v2["Home_Loan"] == 1) |
    (df_products_v2["Personal_Loan"] == 1),
    0.70,
    0.20
)

df_products_v2["Insurance"] = np.random.binomial(
    1, insurance_prob
)

In [23]:
display(df_products_v2.head())
print("Shape:", df_products_v2.shape)

,Customer_ID,Savings_Account,Debit_Card,Credit_Card,Home_Loan,Insurance,Personal_Loan
0,1,1,1,0,0,1,1
1,2,0,0,0,0,0,0
2,3,0,1,0,0,0,0
3,4,1,1,1,1,1,0
4,5,1,1,1,0,1,0


Shape: (29965, 7)


In [24]:
for product in product_columns:
    print(
        product,
        ":",
        df_products_v2[product].sum()
    )

Savings_Account : 21060
Credit_Card : 13024
Debit_Card : 22000
Insurance : 11308
Home_Loan : 4472
Personal_Loan : 7378


In [25]:
import pandas as pd
import numpy as np

# Load the original 30,000-customer product dataset
original_products = pd.read_csv("customer_products.csv")

print("Original dataset shape:", original_products.shape)

n = len(original_products)

df_products_v2 = pd.DataFrame()

df_products_v2["Customer_ID"] = original_products["Customer_ID"]

Original dataset shape: (29965, 7)


In [26]:
np.random.seed(42)

# Savings Account
df_products_v2["Savings_Account"] = np.random.binomial(
    1, 0.70, n
)

# Debit Card - strongly associated with Savings Account
debit_prob = np.where(
    df_products_v2["Savings_Account"] == 1,
    0.90,
    0.35
)

df_products_v2["Debit_Card"] = np.random.binomial(
    1, debit_prob
)

# Home Loan
df_products_v2["Home_Loan"] = np.random.binomial(
    1, 0.15, n
)

# Personal Loan
df_products_v2["Personal_Loan"] = np.random.binomial(
    1, 0.25, n
)

# Insurance - strongly associated with loans
insurance_prob = np.where(
    (df_products_v2["Home_Loan"] == 1) |
    (df_products_v2["Personal_Loan"] == 1),
    0.70,
    0.20
)

df_products_v2["Insurance"] = np.random.binomial(
    1, insurance_prob
)

# Credit Card - strongly associated with Savings + Debit
credit_prob = np.where(
    (df_products_v2["Savings_Account"] == 1) &
    (df_products_v2["Debit_Card"] == 1),
    0.60,
    0.15
)

df_products_v2["Credit_Card"] = np.random.binomial(
    1, credit_prob
)

In [27]:

print("V2 Shape:", df_products_v2.shape)

V2 Shape: (29965, 7)


In [28]:
product_columns = [
    "Savings_Account",
    "Credit_Card",
    "Debit_Card",
    "Insurance",
    "Home_Loan",
    "Personal_Loan"
]

print("Product ownership counts:\n")

for product in product_columns:
    print(
        f"{product}:",
        df_products_v2[product].sum()
    )

Product ownership counts:

Savings_Account: 21060
Credit_Card: 13030
Debit_Card: 22000
Insurance: 11452
Home_Loan: 4483
Personal_Loan: 7461


In [29]:
df_products_v2.to_csv(
    "customer_products_v2.csv",
    index=False
)

print("customer_products_v2.csv saved successfully!")

customer_products_v2.csv saved successfully!


In [30]:
import pandas as pd
import numpy as np

df_products = pd.read_csv("customer_products_v2.csv")

print("Dataset shape:", df_products.shape)
display(df_products.head())

Dataset shape: (29965, 7)


,Customer_ID,Savings_Account,Debit_Card,Home_Loan,Personal_Loan,Insurance,Credit_Card
0,1,1,1,0,0,0,0
1,2,0,0,0,0,0,0
2,3,0,1,0,0,0,0
3,4,1,1,0,1,1,1
4,5,1,1,0,0,1,1


In [31]:
product_columns = [
    "Savings_Account",
    "Credit_Card",
    "Debit_Card",
    "Insurance",
    "Home_Loan",
    "Personal_Loan"
]

In [32]:
transactions_df = df_products[product_columns].astype(bool)

print("Transaction matrix shape:", transactions_df.shape)

display(transactions_df.head())

Transaction matrix shape: (29965, 6)


,Savings_Account,Credit_Card,Debit_Card,Insurance,Home_Loan,Personal_Loan
0,True,False,True,False,False,False
1,False,False,False,False,False,False
2,False,False,True,False,False,False
3,True,True,True,True,False,True
4,True,True,True,True,False,False


In [33]:
transactions_df = transactions_df[
    transactions_df.sum(axis=1) > 0
].copy()

print("Valid customer transactions:", len(transactions_df))

Valid customer transactions: 27455


In [34]:
from mlxtend.frequent_patterns import apriori

frequent_itemsets = apriori(
    transactions_df,
    min_support=0.10,
    use_colnames=True
)

print("Number of frequent itemsets:", len(frequent_itemsets))

Number of frequent itemsets: 32


In [35]:
display(frequent_itemsets.sort_values(
    by="support",
    ascending=False
).reset_index(drop=True))

,support,itemsets
0,0.801311,frozenset({Debit_Card})
1,0.767073,frozenset({Savings_Account})
2,0.689419,"frozenset({Debit_Card, Savings_Account})"
3,0.474595,frozenset({Credit_Card})
4,0.431615,"frozenset({Credit_Card, Debit_Card})"
5,0.425897,"frozenset({Credit_Card, Savings_Account})"
6,0.417119,frozenset({Insurance})
7,0.414715,"frozenset({Credit_Card, Debit_Card, Savings_Ac..."
8,0.304098,"frozenset({Insurance, Debit_Card})"
9,0.291896,"frozenset({Insurance, Savings_Account})"


In [36]:
from mlxtend.frequent_patterns import association_rules

rules = association_rules(
    frequent_itemsets,
    metric="confidence",
    min_threshold=0.50
)

print("Number of association rules:", len(rules))

Number of association rules: 60


In [37]:
rules_display = rules[
    [
        "antecedents",
        "consequents",
        "support",
        "confidence",
        "lift"
    ]
].copy()

rules_display["antecedents"] = rules_display["antecedents"].apply(
    lambda x: ", ".join(sorted(x))
)

rules_display["consequents"] = rules_display["consequents"].apply(
    lambda x: ", ".join(sorted(x))
)

rules_display = rules_display.sort_values(
    by="lift",
    ascending=False
).reset_index(drop=True)

display(rules_display)

,antecedents,consequents,support,confidence,lift
0,"Debit_Card, Personal_Loan","Insurance, Savings_Account",0.118885,0.599779,2.054772
1,"Personal_Loan, Savings_Account","Debit_Card, Insurance",0.118885,0.620532,2.040570
2,Home_Loan,Insurance,0.114333,0.700201,1.678660
3,Personal_Loan,Insurance,0.189073,0.695751,1.667993
4,Personal_Loan,"Debit_Card, Insurance",0.137789,0.507037,1.667348
5,"Debit_Card, Personal_Loan",Insurance,0.137789,0.695149,1.666548
6,"Debit_Card, Personal_Loan, Savings_Account",Insurance,0.118885,0.694173,1.664208
7,"Personal_Loan, Savings_Account",Insurance,0.132690,0.692586,1.660403
8,"Debit_Card, Insurance, Savings_Account",Credit_Card,0.157713,0.602058,1.268572
9,"Credit_Card, Personal_Loan","Debit_Card, Savings_Account",0.102422,0.873835,1.267494


In [38]:
recommendation_rules = rules.copy()
recommendation_rules = recommendation_rules[
    recommendation_rules["consequents"].apply(lambda x: len(x) == 1)
].copy()

In [39]:
recommendation_rules = recommendation_rules[
    (recommendation_rules["confidence"] >= 0.50) &
    (recommendation_rules["lift"] > 1.0) &
    (recommendation_rules["support"] >= 0.10)
].copy()

In [40]:
recommendation_rules = recommendation_rules.sort_values(
    by=["lift", "confidence"],
    ascending=False
).reset_index(drop=True)
print(
    "Number of recommendation rules:",
    len(recommendation_rules)
)

Number of recommendation rules: 36


In [41]:
recommendation_display = recommendation_rules[
    [
        "antecedents",
        "consequents",
        "support",
        "confidence",
        "lift"
    ]
].copy()

recommendation_display["antecedents"] = (
    recommendation_display["antecedents"]
    .apply(lambda x: ", ".join(sorted(x)))
)

recommendation_display["consequents"] = (
    recommendation_display["consequents"]
    .apply(lambda x: ", ".join(sorted(x)))
)

display(recommendation_display)

,antecedents,consequents,support,confidence,lift
0,Home_Loan,Insurance,0.114333,0.700201,1.678660
1,Personal_Loan,Insurance,0.189073,0.695751,1.667993
2,"Debit_Card, Personal_Loan",Insurance,0.137789,0.695149,1.666548
3,"Debit_Card, Personal_Loan, Savings_Account",Insurance,0.118885,0.694173,1.664208
4,"Personal_Loan, Savings_Account",Insurance,0.132690,0.692586,1.660403
5,"Debit_Card, Insurance, Savings_Account",Credit_Card,0.157713,0.602058,1.268572
6,"Debit_Card, Savings_Account",Credit_Card,0.414715,0.601543,1.267487
7,"Credit_Card, Debit_Card, Personal_Loan",Savings_Account,0.102422,0.966988,1.260619
8,"Debit_Card, Personal_Loan, Savings_Account",Credit_Card,0.102422,0.598043,1.260114
9,"Credit_Card, Debit_Card, Insurance",Savings_Account,0.157713,0.961154,1.253015


In [42]:
def recommend_products(
    customer_products,
    rules_df,
    top_n=5
):
    
    recommendations = []

    customer_products = set(customer_products)

    for _, rule in rules_df.iterrows():

        antecedent = set(rule["antecedents"])
        consequent = set(rule["consequents"])

        # Check whether customer owns all antecedent products
        if antecedent.issubset(customer_products):

            # Recommend only products the customer doesn't own
            new_products = consequent - customer_products

            for product in new_products:

                recommendations.append({
                    "Recommended_Product": product,
                    "Confidence": rule["confidence"],
                    "Lift": rule["lift"],
                    "Support": rule["support"]
                })

    if not recommendations:
        return pd.DataFrame(
            columns=[
                "Recommended_Product",
                "Confidence",
                "Lift",
                "Support"
            ]
        )

    recommendations_df = pd.DataFrame(recommendations)

    # Remove duplicate recommendations
    recommendations_df = (
        recommendations_df
        .sort_values(
            by=["Lift", "Confidence"],
            ascending=False
        )
        .drop_duplicates(
            subset=["Recommended_Product"]
        )
        .head(top_n)
        .reset_index(drop=True)
    )

    return recommendations_df

In [43]:
customer = [
    "Savings_Account",
    "Debit_Card"
]

recommendations = recommend_products(
    customer,
    recommendation_rules
)

display(recommendations)

,Recommended_Product,Confidence,Lift,Support
0,Credit_Card,0.601543,1.267487,0.414715


In [44]:
def get_customer_products(row):
    return [
        product
        for product in product_columns
        if row[product]
    ]


customer_recommendations = []

for _, row in df_products.iterrows():

    owned_products = get_customer_products(row)

    recommendations = recommend_products(
        owned_products,
        recommendation_rules,
        top_n=3
    )

    for _, rec in recommendations.iterrows():

        customer_recommendations.append({
            "Customer_ID": row["Customer_ID"],
            "Existing_Products": ", ".join(owned_products),
            "Recommended_Product": rec["Recommended_Product"],
            "Confidence": rec["Confidence"],
            "Lift": rec["Lift"],
            "Support": rec["Support"]
        })

customer_recommendations_df = pd.DataFrame(
    customer_recommendations
)

print(
    "Total recommendations:",
    len(customer_recommendations_df)
)

display(customer_recommendations_df.head(10))

Total recommendations: 22190


,Customer_ID,Existing_Products,Recommended_Product,Confidence,Lift,Support
0,1,"Savings_Account, Debit_Card",Credit_Card,0.601543,1.267487,0.414715
1,3,Debit_Card,Credit_Card,0.538636,1.134939,0.431615
2,3,Debit_Card,Savings_Account,0.860364,1.121618,0.689419
3,9,"Savings_Account, Insurance",Credit_Card,0.556027,1.171583,0.162302
4,9,"Savings_Account, Insurance",Debit_Card,0.898765,1.121618,0.689419
5,13,"Credit_Card, Debit_Card",Savings_Account,0.960844,1.252610,0.414715
6,17,"Savings_Account, Debit_Card, Personal_Loan",Insurance,0.695751,1.667993,0.189073
7,17,"Savings_Account, Debit_Card, Personal_Loan",Credit_Card,0.601543,1.267487,0.414715
8,18,"Savings_Account, Debit_Card, Insurance, Home_Loan",Credit_Card,0.602058,1.268572,0.157713
9,20,"Savings_Account, Debit_Card, Insurance, Person...",Credit_Card,0.602058,1.268572,0.157713


In [45]:
customer_recommendations_df.to_csv(
    "customer_cross_sell_recommendations.csv",
    index=False
)

print("Cross-sell recommendations saved successfully!")

Cross-sell recommendations saved successfully!


# Phase 9 — Cross-Sell Recommendation System

## Objective

The objective of this phase was to build a banking cross-sell recommendation system using customer product ownership data and association rule mining.

## 1. Product Dataset

A synthetic banking product dataset was created containing:

- Customer_ID
- Savings_Account
- Credit_Card
- Debit_Card
- Insurance
- Home_Loan
- Personal_Loan

The final product dataset contained **29,965 customers** and **6 banking products**.

## 2. Transaction Representation

Each customer's product ownership was represented as a transaction.

Example:

```text
Customer:
Savings Account + Debit Card + Insurance